File: demo/java/Point3d.java
import java.io.*;
public class Point3d{
public float x;
public float y;
public float z;
public void serialize(DataOutputStream dos) throws Exception{
FieldMark fm = new FieldMark(1);
fm.mark(x == 0);
fm.mark(y == 0);
fm.mark(z == 0);
if(fm.isMarked(0)){
dos.writeFloat(x);
}
if(fm.isMarked(1)){
dos.writeFloat(y);
}
if(fm.isMarked(2)){
dos.writeFloat(z);
}
}
public void deserialize(DataInputStream dis) throws Exception{
FieldMark fm = new FieldMark(1);
dis.read(fm.getData());
if(fm.readMark()){
x = dis.readFloat();
}
if(fm.readMark()){
y = dis.readFloat();
}
if(fm.readMark()){
z = dis.readFloat();
}
}
}

File: demo/java/src/main/java/info/chenliang/tank/Point3d.cs
using System;
using System.IO;
using MiscUtil;
using MiscUtil.IO;
using MiscUtil.Conversion;
using Talky;
namespace info.chenliang.tank{
public class Point3d{
public single x;
public single y;
public single z;
public void Serialize(MemoryEndianBinaryWriter writer){
FieldMark fm = new FieldMark(1);
fm.Mark(x != 0);
fm.Mark(y != 0);
fm.Mark(z != 0);
writer.Write(fm.getData());
if(fm.IsMarked(0)){
writer.WriteSingle(x);
}
if(fm.IsMarked(1)){
writer.WriteSingle(y);
}
if(fm.IsMarked(2)){
writer.WriteSingle(z);
}
}
public void Deserialize(MemoryEndianBinaryReader reader){
FieldMark fm = new FieldMark(1);
reader.Read(fm.getData());
if(fm.ReadMark()){
x = reader.ReadSingle();
}
if(fm.ReadMark()){
y = reader.ReadSingle();
}
if(fm.ReadMark()){
z = reader.ReadSingle();
}
}
}
}
